Navigation (online&cached) added

quaso opened the pull request.

Commit:
mazda internal patch enhanced with HTML5 geo location

quaso committed:
test env added

quaso committed:
sandbox for graphhopper created

quaso committed:
requesting from graphhopper is working

quaso committed:
parsing and setting route

quaso committed:
minor fix

quaso committed:
minor fix2

quaso committed:
showing turn and distance-to-turn as driving via the computed route (rerouting is not implemented yet)

quaso committed:
- fixed an issue with two consecutive CONTINUE_ON_STREET instructions
- code cleanup

quaso committed:
displaying route with marker icons

quaso committed:
error handling improved

quaso committed:
rerouting

quaso committed:
- rerouting fixed
- off route indicators added

quaso committed:
several continue instructions merged to next non-continue one

Commit:
showing line instead of points for the route

quaso committed:
destination reached fixed

quaso committed:
maxOffRouteTime removed from settings

quaso committed:
- menu manager moved to extra file menu.js (currently just mainMenu is supported)
- javascripts loading improved

Commit:
Menu manager created, multiple menus are supported

Commit:
dynamically creating list of destinations in navigation menu

Commit:
NavCtrl.css formatting fixed

Commit:
$.ajax() error handling improved

Commit:
caching routes and reading routes from cache

Commit:
readme updated

Commit:
GraphHopper attribution shown

Commit:
testing improved

Commit:
readme updated

flyandi commented:
Howdy - Thanks for the pull request. I just skimmed through your additions and it looks pretty good. I will test drive your code (literally).

Couple questions before I merge the request:

1) Did you split out my menu maker?
2) Maybe I missed it, but I didn't see anything related to input the address.
3) You said over in the forums that the route recalculates itself every 5seconds - will that count against the API? 
4) I may want to reorganize some of your code, so we can better abstract from that. I was playing around with Google's routing API and had a similar implementation. Would be nice to get to a point to support different providers.

Anyway, I will test this weekend :-)

Good job!

quaso commented:
Hi
1. Yes, I moved menu into two separate js files. One contains menu data and the other contains list of menus. I was annoyed with one long menu, with your items and then items with navigation. So I created two menus and implemented a logic for having multiple independent menus.
2. Input address is not implemented yet. One has to hardcode destination GPS coordinates into settings.js file. I do not like it, but it is quite fine for the beginning. I am planning to create some virtual keyboard (or use some existing one), so the user can enter the address. Graphhopper has even geocoding API, so I will use it.
3. The route is recalculated only if you are off route. So when you are driving according to the route (+/- 30 metres), then there are no POST request. And yes, every route (re)calculation takes one credit at Graphhopper.
4. I wasnt playing with other routing providers, because ffwdme project was using Graphhopper already. One of the great advatnages of this service is, that they provide you list of points. These points represents GPS coorditanes for the route. If you have one long road, then you receive just two points. But if you have curvy road, you receive tens of points. So I can use all these points in OpenLayers LineString to display the route. I do not know, if any other service provides this. However, I do not want to stick to one service provider only. That is why routing logic is based on info parsed from Graphhoper response (see parse method in graph_hopper.js). Actually it is only list of distances, turn types, and points at the route. I parse out instruction text and duration as well, but these are not used. So any service provider, which provides those three info can be used.

flyandi commented:
The system already has a full keyboard instance that is used for the SMS and Email input. It's a simple object that is globally available. I will work on that part once I merged the pull request.

What's the ffwdme project?

quaso commented:
https://github.com/ffwdme/ffwdme.js - navigation in javascript. It has many features, which I did not included into this project.

quaso committed:
algorithm for parsing data from Graphhopper rewritten, because of wrong results of the original (graphhopper instruction contains turn and then distance/points to the next turn)

quaso committed:
Showing turn instruction text in the notificationDiv

quaso committed:
exporting&clearing localStorage to web app

quaso committed:
UI minor fixes

quaso committed:
Java project for editing cached routes added

quaso committed:
URI for routes manager is configurable

quaso committed:
distribution package updated

Commit:
- controlNotification inserted into ctrlMap, was in ctrlDiv
- route color is configurable

Commit:
transforming routes from/to json and zip implemented

Commit:
uplading file from javascript web app to Java web app

quaso committed:
displaying uploaded routes in a list

Commit:
UI fixes to notificationCtrl

Commit:
Merge branch 'brettalcox-fixUI' into fixUI (CSS fixes of notification
panel)

Commit:
fixes in showNotification function

Commit:
- showing a route when user clicks on it
- adding Bootstrap navigation bar

IgorKvasn committed:
ui fixes

Commit:
route cache management added to separate js file

Commit:
showing and clearing more routes implemented

Commit:
offline navigation with cached routes

quaso committed:
debug messages for offline routing shown

Commit:
visibility css added

Commit:
debug mode for offline routing

Commit:
showing more notifications in offline routing mode

Commit:
rounding numbers in LatLng (fixing float arithmetic)

quaso committed:
offline navigation tweaking & DestinationHolder created

Commit:
minor bug fix

Commit:
minor bug fixes

quaso committed:
- no connection icon added
- off route icons changed

quaso committed:
Merge branch 'master' into export

quaso committed:
- start/finish marker icons added and shown
- mapView starts with current user locaction

quaso committed:
adding new route (not storing it yet)

quaso committed:
uploaded to Google App Engine & running there :)

quaso committed:
sending exported routes via email

quaso committed:
File uploading fixed

quaso committed:
loading api key and other settings from settings.js (user must upload it)

quaso committed:
loading destinations from settings.js file (and showing them in context menu)

quaso committed:
adding new route

quaso committed:
saving routes to files (not finished yet)

quaso committed:
showing alerts on map (loaded from alerts.js)

quaso committed:
Parsing KML file with alerts and creating alerts.js

quaso committed:
typo error fixed

flyandi commented:
Hey quaso,

I had finally a chance to test the code tonight however when I load the files up, it won't show up any tiles. 

Is there anything else I have to do to get this working?

flyandi commented:
also used your latest repo, btw.

brettalcox commented:
I attached a .zip of the test.css, test.html, test.js. Put these in the /src/ directory (where jci, jci-patch, system are), and put the routesCacheFile.js in there as well. Then the settings.js file goes in the /src/system/js folder. There is a slight difference in the settings file I attached and the one that is in the latest branch (a comma needed to be added).

I currently use the test.html to input a location, have it generate the route, then go into the browser localStorage and grab the long string of directions/coordinates to put in the routesCacheFile (and then add the destination in the settings.js). But if you have constant internet while driving, all you need to do is put destination in settings file.

[src.zip](https://github.com/flyandi/mazda-enhanced-compass/files/115722/src.zip)

quaso commented:
At first I did not notice your test env, so I created my own. As it turned out lately, my changes in code are not compatible with your test env, but at least they are working in real car :). In near future I will put some effort into adjusting the code, so it will work in your test env and I can remove mine. Till that time, please use workaround and zip posted by brettalcox.

Commit:
downloading routes as zip

Commit:
added example for alerts.js

Commit:
adjusted for flyandi's test environment

quaso commented:
It was not so hard to fix actually :) There were some relative path issues. Please try now

flyandi commented:
Ok - I will retry tonight. Thanks guys.

quaso committed:
minor fix for clearing routes in session after zip file was created

Commit:
syncing route format with one used in Mazda app

Commit:
do not using CDN for fontawesome and jquery

Commit:
CORS issue fixed (Spring boot version upgrade)

Commit:
run as SpringBoot app or in GAE

Commit:
destination editor added

quaso committed:
creating destination improved

Commit:
- cleaning new route and start/finish markers fixed
- openlayers library updated

quaso committed:
Saving settings.js in UI

quaso committed:
Merge branch 'export'

quaso committed:
entering GraphHopper API key improved

Commit:
- fixed exporting routes with asynchronous upload
- webUI is able to edit locale and route color

Commit:
File uploading in webUI improved

quaso committed:
CORS issue fixed

Commit:
lazy loading alerts.js

Commit:
Message is displayed when there are no routes for export

kkaczmarczyk commented:
Was this tested in the end?;)

quaso commented:
I tested it, but it was a long time ago :)
